server/lib/scheduler/Scheduler.ts

```diff
@@ -20,67 +31,102 @@ class Scheduler {
   }
 
   public async assignNextWorkspace(userId) {
-    // clear cache so we don't rely on old rootParent information
+    // clear caches so we don't rely on old information
     this.rootParentCache.clearRootParentCache();
+    this.remainingBudgetAmongDescendantsCache.clearRemainingBudgetAmongDescendantsCache();
+
+    const actionableWorkspaces = await this.getActionableWorkspaces(userId);
 
-    const actionableWorkspaces = await this.getActionableWorkspaces();
-    console.log("actionableWorkspaces", actionableWorkspaces);
-    if (actionableWorkspaces.length === 0) {
-      return;
-    }
     const assignedWorkspace = pickRandomItemFromArray(actionableWorkspaces);
-    console.log("assignedWorkspace", assignedWorkspace && assignedWorkspace.id);
+
+    if (!assignedWorkspace) {
+      throw new Error("No workspace to choose from");
+    }
+
     await this.schedule.assignWorkspaceToUser(userId, assignedWorkspace);
   }
 
-  private async getActionableWorkspaces() {
-    const workspacesInTreesWorkedOnLeastRecently = await this.getWorkspacesInTreesWorkedOnLeastRecently();
-    const eligibleWorkspaces = await this.filterByEligibility(workspacesInTreesWorkedOnLeastRecently);
-    const notYetWorkedOn = await this.filterByWhetherNotYetWorkedOn(eligibleWorkspaces);
+  private async getActionableWorkspaces(userId) {
+    let treesToConsider = await this.fetchAllRootWorkspaces();
+
+    while (treesToConsider.length > 0) {
+      const leastRecentlyWorkedOnTreesToConsider = await this.getTreesWorkedOnLeastRecentlyByUser(userId, treesToConsider);
+      const randomlySelectedTree = pickRandomItemFromArray(leastRecentlyWorkedOnTreesToConsider);
+      const actionableWorkspaces = await this.getActionableWorkspacesForTree(userId, randomlySelectedTree);
+
+      if (actionableWorkspaces.length > 0) {
+        return actionableWorkspaces;
+      } else {
+        treesToConsider = _.difference(
+          treesToConsider,
+          [randomlySelectedTree]
+        );
+      }
+    }
+  }
```

Comment:
There must be some better ways of writing the above code, for readability. I'd want something that was more declarative, I found this pretty confusing.